src/main/java/com/griefprevention/protection/ProtectionHelper.java

```diff
@@ -0,0 +1,108 @@
+package com.griefprevention.protection;
+
+import me.ryanhamshire.GriefPrevention.Claim;
+import me.ryanhamshire.GriefPrevention.ClaimPermission;
+import me.ryanhamshire.GriefPrevention.ClaimsMode;
+import me.ryanhamshire.GriefPrevention.DataStore;
+import me.ryanhamshire.GriefPrevention.GriefPrevention;
+import me.ryanhamshire.GriefPrevention.Messages;
+import me.ryanhamshire.GriefPrevention.PlayerData;
+import me.ryanhamshire.GriefPrevention.events.PreventBlockBreakEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.Material;
+import org.bukkit.World;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Event;
+import org.bukkit.event.block.BlockBreakEvent;
+import org.bukkit.event.block.BlockPlaceEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+
+/**
+ * A utility used to simplify various protection-related checks.
+ */
+public final class ProtectionHelper
+{
+
+    private ProtectionHelper() {}
+
+    /**
+     * Check the {@link ClaimPermission} state for a {@link Player} at a particular {@link Location}.
+     *
+     * <p>This respects ignoring claims, wilderness rules, etc.</p>
+     *
+     * @param player the person performing the action
+     * @param location the affected {@link Location}
+     * @param permission the required permission
+     * @param trigger the triggering {@link Event}, if any
+     * @return the denial message supplier, or {@code null} if the action is not denied
+     */
+    public static @Nullable Supplier<String> checkPermission(
+            @NotNull Player player,
+            @NotNull Location location,
+            @NotNull ClaimPermission permission,
+            @Nullable Event trigger)
+    {
+        World world = location.getWorld();
+        if (world == null || !GriefPrevention.instance.claimsEnabledForWorld(world)) return null;
+
+        PlayerData playerData = GriefPrevention.instance.dataStore.getPlayerData(player.getUniqueId());
+
+        // Administrators ignoring claims always have permission.
+        if (playerData.ignoreClaims) return null;
+
+        Claim claim = GriefPrevention.instance.dataStore.getClaimAt(location, false, playerData.lastClaim);
+
+
+        // If there is no claim here, use wilderness rules.
+        if (claim == null)
+        {
+            ClaimsMode mode = GriefPrevention.instance.config_claims_worldModes.get(world);
+            if (mode == ClaimsMode.Creative || mode == ClaimsMode.SurvivalRequiringClaims)
```

Comment:
I'm thinking of removing these claim modes. Guess it doesn't have to happen now, but I think it is something I'll do for v17.

Reply:
Good by me. I would still like a utility for claim stuff either way, members like `EntityEventHandler#isBlockSourceInClaim(ProjectileSource, Claim)` are building up in various places and I'd really love to unify entity -> player source determination somewhere if possible.